src/java/org/apache/cassandra/locator/Replica.java

```
@@ -0,0 +1,221 @@
+    public ReplicaSet subtractByRange(Replicas toSubtract)
+    {
+        if (isFull() && Iterables.all(toSubtract, Replica::isFull))
+        {
+            Set<Range<Token>> subtractedRanges = getRange().subtractAll(toSubtract.asRangeSet());
+            ReplicaSet replicaSet = new ReplicaSet(subtractedRanges.size());
+            for (Range<Token> range: subtractedRanges)
+            {
+                replicaSet.add(new Replica(getEndpoint(), range, isFull()));
+            }
+            return replicaSet;
+        }
+        else
+        {
+            // FIXME: add support for transient replicas
+            throw new UnsupportedOperationException("transient replicas are currently unsupported");
+        }
+    }
+
+    public ReplicaList normalizeByRange()
+    {
+        List<Range<Token>> normalized = Range.normalize(Collections.singleton(getRange()));
+        ReplicaList replicas = new ReplicaList(normalized.size());
+        for (Range<Token> normalizedRange: normalized)
+        {
+            replicas.add(new Replica(getEndpoint(), normalizedRange, isFull()));
+        }
+        return replicas;
+    }
+
+    public boolean contains(Range<Token> that)
+    {
+        return getRange().contains(that);
+    }
+
+    public boolean intersectsOnRange(Replica replica)
+    {
+        return getRange().intersects(replica.getRange());
+    }
+
+    public Replica decorateSubrange(Range<Token> subrange)
+    {
+        Preconditions.checkArgument(range.contains(subrange));
+        return new Replica(getEndpoint(), subrange, isFull());
+    }
+
+    public static Replica full(InetAddressAndPort endpoint, Range<Token> range)
+    {
+        return new Replica(endpoint, range, true);
+    }
+
+    /**
+     * We need to assume an endpoint is a full replica in a with unknown ranges in a
+     * few cases, so this returns one that throw an exception if you try to get it's range
+     */
+    public static Replica fullStandin(InetAddressAndPort endpoint)
```

This feels so unsafe. At the first level it's turning a compile time error into a runtime error. But then it's turning the fullness or not of the replica into a silent error. Do things work if you have isFull and isTransient throw as well?

Another issue is this is going to make isFull() and isTransient() bi-morphic which isn't the best. Feels like it might be premature optimization to worry about that right now though.

---

I agree it sucks. The problem is that there batchlog and hint related functions that don't work particularly well with the Replica idea, but use the snitch and some write path stuff that were converted to use Replicas to prevent a lot of conversion/copying on the hot path. So using fake replicas solved that problem, but not that well. Let me take another look at the places that use stand ins though, there's probably something less crappy that can be done.